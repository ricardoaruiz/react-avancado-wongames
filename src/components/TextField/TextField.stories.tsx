import { Story, Meta } from '@storybook/react/types-6-0'
import styled from 'styled-components'
import { Email } from '@styled-icons/material-outlined/Email'

import TextField, { TextFieldProps } from './TextField'

export default {
  title: 'TextField',
  component: TextField,
  argTypes: {
    value: {
      type: 'string'
    },
    iconPosition: {
      defaultValue: 'left',
      control: {
        type: 'select',
        options: ['left', 'right']
      }
    },
    icon: {
      type: ''
    },
    onChange: {
      action: 'change'
    }
  }
} as Meta

const Container = styled.div`
  max-width: 30rem;
  margin: 0 auto;
`

export const Basic: Story<TextFieldProps> = (args) => (
  <Container>
    <TextField {...args} />
  </Container>
)

Basic.args = {
  label: 'Email',
  labelFor: 'email',
  placeholder: 'Email',
  value: ''
}

export const WithIcon: Story<TextFieldProps> = (args) => (
  <Container>
    <TextField {...args} />
  </Container>
)

WithIcon.args = {
  label: 'Email',
  labelFor: 'email',
  placeholder: 'Email',
  value: '',
  icon: <Email />
}

export const WithError: Story<TextFieldProps> = (args) => (
  <Container>
    <TextField {...args} />
  </Container>
)

WithError.args = {
  label: 'Email',
  labelFor: 'email',
  placeholder: 'Email',
  value: '',
  icon: <Email />,
  error: 'Required field'
}
