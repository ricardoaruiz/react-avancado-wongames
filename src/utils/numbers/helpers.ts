export const toCurrencyString = (value: number) =>
  Intl.NumberFormat('pt-BR', {
    style: 'currency',
    currency: 'BRL'
  }).format(value)
