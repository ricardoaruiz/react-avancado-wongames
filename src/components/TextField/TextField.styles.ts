import styled, { css, DefaultTheme } from 'styled-components'

import { TextFieldProps } from './TextField'

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    margin-bottom: ${theme.spacings.xxsmall};
    cursor: pointer;
  `};
`
type WrapperInputProps = Pick<TextFieldProps, 'iconPosition'>

const wrapperInputMofiers = {
  iconPosition: {
    left: (theme: DefaultTheme) => css`
      ${Input} {
        order: 1;
      }
      ${Icon} {
        order: 0;
        margin-right: ${theme.spacings.xxsmall};
      }
    `,
    right: (theme: DefaultTheme) => css`
      ${Input} {
        order: 0;
      }
      ${Icon} {
        order: 1;
        margin-left: ${theme.spacings.xsmall};
      }
    `
  }
}

export const WrapperInput = styled.div<WrapperInputProps>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    align-items: center;
    justify-content: space-between;

    width: 100%;
    border-radius: ${theme.border.radius};
    background-color: ${theme.colors.lightGray};
    border-radius: ${theme.border.radius};
    padding: 0 ${theme.spacings.xxsmall};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    ${wrapperInputMofiers.iconPosition[iconPosition!](theme)};
  `};
`

export const Input = styled.input`
  ${({ theme }) => css`
    flex: 1;
    width: inherit;
    height: 100%;
    border: none;
    outline: none;
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    background-color: transparent;
    padding: ${theme.spacings.xxsmall} 0;
  `};
`

export const Icon = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: center;

    & svg {
      width: 2.2rem;
      color: ${theme.colors.gray};
    }
  `};
`

type WrapperProps = Pick<TextFieldProps, 'disabled'>

const wrapperModifier = {
  disabled: (theme: DefaultTheme) => css`
    ${Input},
    ${Label} {
      color: ${theme.colors.gray};
      cursor: default;
    }
  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, disabled }) => css`
    ${disabled && wrapperModifier.disabled(theme)}
  `};
`
