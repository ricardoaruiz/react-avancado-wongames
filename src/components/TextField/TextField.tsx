import React, { InputHTMLAttributes } from 'react'

import * as S from './TextField.styles'

export type TextFieldIconPositions = 'left' | 'right'

export type TextFieldProps = {
  label?: string
  placeholder?: string
  value?: string
  icon?: React.ReactNode
  iconPosition?: TextFieldIconPositions
  disabled?: boolean
  error?: string
  onValueChange?: (value: string) => void
} & InputHTMLAttributes<HTMLInputElement>

export const TextField = ({
  label,
  name,
  placeholder,
  value = '',
  icon,
  iconPosition = 'left',
  disabled = false,
  error = '',
  onValueChange,
  ...props
}: TextFieldProps) => {
  const [inputValue, setInputValue] = React.useState(value)

  const onValueChangeHandler = React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const value = e.currentTarget.value
      setInputValue(value)
      !!onValueChange && onValueChange(value)
    },
    [onValueChange]
  )

  return (
    <S.Wrapper disabled={disabled} hasError={!!error}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}

      <S.WrapperInput iconPosition={iconPosition}>
        {!!icon && <S.Icon>{icon}</S.Icon>}

        <S.Input
          type="text"
          placeholder={placeholder}
          value={inputValue}
          onChange={onValueChangeHandler}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </S.WrapperInput>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}

export default TextField
