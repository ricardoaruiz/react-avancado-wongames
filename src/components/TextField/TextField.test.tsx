import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderWithTheme } from 'utils/tests/helpers'
import { Email } from '@styled-icons/material-outlined/Email'
import theme from 'styles/theme'

import TextField from './TextField'

describe('<TextField />', () => {
  it('should render without label', () => {
    const { container } = renderWithTheme(<TextField />)

    expect(screen.getByRole('textbox')).toBeInTheDocument()
    expect(container.querySelector('label')).not.toBeInTheDocument()
  })

  it('should render with label', () => {
    renderWithTheme(<TextField label="Name" />)

    expect(screen.getByText(/name/i)).toBeInTheDocument()
  })

  it('should render with placeholder', () => {
    renderWithTheme(<TextField placeholder="Name" />)

    expect(screen.getByPlaceholderText(/name/i)).toBeInTheDocument()
  })

  it('should render with default value', () => {
    renderWithTheme(<TextField value="test@email.com" />)

    expect(screen.getByRole('textbox')).toHaveAttribute(
      'value',
      'test@email.com'
    )
  })

  it('should call onValueChange when value is typed', async () => {
    const onValueChange = jest.fn()
    const text = 'test@email.com'
    renderWithTheme(<TextField onValueChange={onValueChange} />)

    const inputText = screen.getByRole('textbox')
    userEvent.type(inputText, text)

    await waitFor(() => {
      expect(inputText).toHaveValue(text)
      expect(onValueChange).toHaveBeenCalledTimes('test@email.com'.length)
      expect(onValueChange).toHaveBeenCalledWith(text)
    })
  })

  it('should be acessible using tab', () => {
    renderWithTheme(<TextField />)

    expect(document.body).toHaveFocus()
    userEvent.tab()
    expect(screen.getByRole('textbox')).toHaveFocus()
  })

  it('should render with left icon by default', () => {
    renderWithTheme(<TextField icon={<Email />} />)

    const input = screen.getByRole('textbox')
    expect(input).toHaveStyle({
      order: 1
    })
    const wrapperInput = input.parentElement
    expect(wrapperInput?.querySelector('div')).toHaveStyle({
      order: 0
    })
    expect(wrapperInput?.querySelector('svg')).toBeInTheDocument()
  })

  it('should render with right icon when iconPosition is passed with right', () => {
    renderWithTheme(<TextField icon={<Email />} iconPosition="right" />)

    const input = screen.getByRole('textbox')
    expect(input).toHaveStyle({
      order: 0
    })
    const wrapperInput = input.parentElement
    expect(wrapperInput?.querySelector('div')).toHaveStyle({
      order: 1
    })
    expect(wrapperInput?.querySelector('svg')).toBeInTheDocument()
  })

  it('should render disable input when disabled is passed', () => {
    renderWithTheme(<TextField label="Email" labelFor="email" disabled />)

    expect(screen.getByRole('textbox')).toBeDisabled()
    expect(screen.getByRole('textbox')).toHaveStyle({
      color: theme.colors.gray,
      cursor: 'default'
    })
    expect(screen.getByText(/email/i)).toHaveStyle({
      color: theme.colors.gray,
      cursor: 'default'
    })
  })

  it('should render error message when is passed', () => {
    const { container } = renderWithTheme(
      <TextField
        label="Email"
        labelFor="email"
        error="Required field"
        icon={<Email data-testid="icon" />}
      />
    )

    const error = screen.getByText(/required field/i)
    expect(error).toBeInTheDocument()
    expect(error).toHaveStyle({
      color: theme.colors.primary
    })
    expect(container.querySelectorAll('div')[1]).toHaveStyle({
      border: '1px solid #F231A5'
    })
    expect(screen.getByTestId('icon').parentElement).toHaveStyle({
      color: theme.colors.primary
    })

    expect(container.firstChild).toMatchSnapshot()
  })
})
