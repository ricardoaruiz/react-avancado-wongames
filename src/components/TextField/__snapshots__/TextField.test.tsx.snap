// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render error message when is passed 1`] = `
.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
  margin-bottom: 0.8rem;
  cursor: pointer;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  border: 1px solid #EAEAEA;
  border-radius: 0.4rem;
  background-color: #EAEAEA;
  border-radius: 0.4rem;
  padding: 0 0.8rem;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c4 .c8 {
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
}

.c4 .c5 {
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  margin-right: 0.8rem;
}

.c9 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: inherit;
  height: 100%;
  border: none;
  outline: none;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  background-color: transparent;
  padding: 0.8rem 0;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #8F8F8F;
}

.c6 svg {
  width: 2.2rem;
}

.c10 {
  display: block;
  margin-top: 0.2rem;
  font-size: 1rem;
  color: #F231A5;
}

.c0 .c1,
.c0 .c5 {
  color: #F231A5;
}

.c0 .c3 {
  border: 1px solid #F231A5;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="email"
  >
    Email
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5 c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        data-testid="icon"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M22 6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6zm-2 0l-8 5-8-5h16zm0 12H4V8l8 5 8-5v10z"
        />
      </svg>
    </div>
    <input
      class="c8 c9"
      id="email"
      type="text"
      value=""
    />
  </div>
  <small
    class="c10"
  >
    Required field
  </small>
</div>
`
